source/base/ByteBuffer.ooc

```diff
@@ -15,21 +16,24 @@ ByteBuffer: class {
 	_size: Int
 	_referenceCount: ReferenceCounter
 	_ownsMemory: Bool
+	_allocator: AbstractAllocator = null
 	pointer ::= this _pointer
 	size ::= this _size
 	referenceCount ::= this _referenceCount
 
-	init: func (=_pointer, =_size, ownsMemory := false) {
+	init: func (=_pointer, =_size, ownsMemory := false, allocator := null as AbstractAllocator) {
 		this _referenceCount = ReferenceCounter new(this)
 		this _ownsMemory = ownsMemory
+		this _allocator = allocator ?? MallocAllocator new() as AbstractAllocator
 	}
 	free: override func {
 		if (this _referenceCount != null)
 			this _referenceCount free()
 		this _referenceCount = null
 		if (this _ownsMemory)
-			memfree(this _pointer)
+			this _allocator deallocate(this _pointer)
 		this _pointer = null
+		this _allocator free()
```

Comment:
It's not uncommon to have an allocator as a singleton for example. In most cases you will maybe have a few different allocators, and to allocate and free these for every ByteBuffer seems like a waste. We could for example have a few static standard allocators in the ByteBuffer class.

```ooc
buffer := ByteBuffer new(size, Allocator malloc)
```